### PR TITLE
out-of-slice rendering for Vectors

### PR DESCRIPTION
**Pull request by brisvag**
# Description
This PR is trying to implement #2832.

https://user-images.githubusercontent.com/23482191/122985586-a87d2780-d39e-11eb-98b4-a7ffff8a7d27.mp4

`Vectors` layer with the `n_dimensional` flag on will be rendered out-of-slice just like Points.
- a vector is rendered based on whether the n_dim vector would intersect with the currently viewed slice.
- the out-of-sliceness of the vector is conveyed by reducing the opacity of the vector from 1 (in slice) to 0 (out of slice). There might be better ways to do this, but this was the first one I tried (see #2832)
- I copied GUI stuff from the `Points` layer, so this is also settable with a checkbox and events should be hooked up correctly

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# References
#2832

# How has this been tested?
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.

### CODE REVIEW

**Review by alisterburt — CHANGES_REQUESTED** [review]
@brisvag this is so cool!!  

I had a quick play with it and there are a few things I'll bring up before delving into the code
- this looks absolutely beautiful when you have a bunch of densely packed vectors around a sphere

https://user-images.githubusercontent.com/7307488/123006258-4bd63880-d3af-11eb-949e-1c113725ce59.mp4

- this PR as is seems to have an effect on the visibility of vectors even when n-dimensional is not set to true on the vectors layer, do you have any ideas why? (left is master, right is this PR)
image

- the transparency/rendering only seems to work when the vectors are projected in one direction. For this fully packed sphere I only see one half of the vectors on the sphere rendered, any ideas what's going on here?

https://user-images.githubusercontent.com/7307488/123006492-a2437700-d3af-11eb-849d-9b49df379c52.mp4

- in the nD vectors example, viewed from the side the transparency effect looks a little strange - is this an extreme example where projecting the whole length of the vector is not quite what we want? (I know it's much harder to change the length/project only a portion)
image

Overall this is a feature I think we should definitely push for! 🚀 I'll take a look at the code in detail in a second pass soon. Adding a comment in #2832 about the unification of this rendering with the behaviour of points/shapes 🙂

**Comment by brisvag** [comment]
I think there are several problems with this... Apart from the bugs above (definitely must have messed up something in the slicing) there are a few things that make me think this is not the right approach.

As it is now, vectors are rendered when out of plane if they intersect the currently viewed slice, and their transparency is based on how far along the vector this intersection happens (PS: at least, this was the intent, I probably messed up somewhere). One problem with this is that vectors that are fully in plane will appear and disappear insantly, as opposed to a vector almost perpendicular to the plane, which will stay around for a while. While maybe it's better than the current master, I also find it a bit counterintuitive.

Another thing is what we were mentioning in #2832 about unifying this behaviour. We would need a unified system to decide how "thick" the object is, and how to convey its out-of-planeness.

Ultimately, I feel like the best way would be to set a slice thickness and simply use that to decide what to display and what not... I'm sure it sounds familiar, @alisterburt :P This actually sounds quite doable on this side, but It's probably supposed to be on the vispy side of things...

**Comment by sofroniewn** [comment]
@brisvag just checking this one out, the videos looks great! I can spend some time over the next week thinking about this one, but wanted to hear if you or @alisterburt have any more recent thoughts/ ideas before I do. Will be fun to get this merged!

**Comment by brisvag** [comment]
I'm not sure where to go with this. While I think it's a useful feature, I also think this is not "the right way" to do it (a range-slider for the dimensions seems to be the mor eintuitive solution). On the other hand, I think this should be fixable and mergeable without too much work, so I can take a look in the coming days!

**Comment by brisvag** [comment]
Came back to this after some time :P Despite what said above, I think it's probably useful to get this feature in as is. It does the job, though it has odd effects in some cases, as shown in the example. some of them are however fixable.

> this PR as is seems to have an effect on the visibility of vectors even when n-dimensional is not set to true on the vectors layer, do you have any ideas why? (left is master, right is this PR)

I tracked down the issue: when `n_dimensional` is disabled, `Points` uses `distance < 1e-5` to determine whether a point is on the slice. I copied this behaviour in `Vectors` on this PR. However, this differs from the current `Vectors` on main, which instead simply converts the positions with `astype(int)`. This is kinda like doing `distance < 1`, except it's always rounded down (which is definitely worse).

The way I see it, we should change *both* of them to use `distance <= 0.5`; this way, we ensure that they are not lost between slices (as happens for points currently). The downside of this is obviously that we are less true to the data. What do other people think? I'm sure @jni and @sofroniewn have opinions on this :)

> the transparency/rendering only seems to work when the vectors are projected in one direction. For this fully packed sphere I only see one half of the vectors on the sphere rendered, any ideas what's going on here?

Fixed, was just a missing `abs`!

> in the nD vectors example, viewed from the side the transparency effect looks a little strange - is this an extreme example where projecting the whole length of the vector is not quite what we want? (I know it's much harder to change the length/project only a portion)

This is an intrinsic issue with this method: we determine whether to show a vector based on whether it *passes* through the slice. This means that vectors parallel to the slice will be visible in a single slice, while something perpendicular will be visible for many slices (albeit more transparent the more distant from its origin). To get something more intuitive, you just need to display a thick plane in 3D, which is a whole other thing we should add for points/vectors just like @alisterburt did for volumes :)


Here's how it looks now:

https://user-images.githubusercontent.com/23482191/150379968-36505bbc-1dc5-479c-bd8b-085af6f29c1c.mp4

As opposed to disabled (at the end I show how it is in 3D):

https://user-images.githubusercontent.com/23482191/150380039-0872c13e-8d95-44ac-90e9-e1e8d868d606.mp4

File: napari/layers/vectors/vectors.py [hunk 1]
```diff
@@ -593,15 +617,64 @@ def edge_contrast_limits(
...
+            Indices to slice with.
+
+        Returns
+        -------
+        slice_indices : list
+            Indices of vectors in the currently viewed slice.
+        alpha : float, (N, ) array
+            If in `n_dimensional` mode then the opacity of vectors, where
+            values of 1 corresponds to vectors located in the slice, and values
+            less than 1 correspond to vectors located in neighboring slices.
```
Comment (alisterburt) [inline napari/layers/vectors/vectors.py]: could you add here whether the alpha values are for the vectors currently in view or the whole set of vectors?

File: napari/layers/vectors/vectors.py [hunk 2]
```diff
@@ -593,15 +617,64 @@ def edge_contrast_limits(
...
+        ----------
+        dims_indices : sequence of int or slice
+            Indices to slice with.
+
+        Returns
+        -------
+        slice_indices : list
+            Indices of vectors in the currently viewed slice.
+        alpha : float, (N, ) array
+            If in `n_dimensional` mode then the opacity of vectors, where
```
Comment (alisterburt) [inline napari/layers/vectors/vectors.py]:
```suggestion
            If in `n_dimensional` mode then the opacity of vectors in the current slice, where
```

will probably break formatting if merged as is

**Review by alisterburt — COMMENTED** [review]
nice! great spelunking for the last few little issues and the results are looking amazing!

We should probably add tests for the new vectors ndim checkbox, checking that updating the view updates the model and vice versa, otherwise this looks good to go - only one minor docstring suggestion from me

> The way I see it, we should change both of them to use distance <= 0.5; this way, we ensure that they are not lost between slices (as happens for points currently). The downside of this is obviously that we are less true to the data. What do other people think? I'm sure @jni and @sofroniewn have opinions on this :)

This seems more intuitive and lots of people are confused when their floating point 3+d points disappear between slider steps, I'm supportive but happy to defer to others if they don't like it

File: napari/layers/vectors/vectors.py (hunk 1 above)
Comment (brisvag) [reply napari/layers/vectors/vectors.py]: I rewrote the dosctring, hopefully now it's more clear.

**Comment by brisvag** [comment]
> This seems more intuitive and lots of people are confused when their floating point 3+d points disappear between slider steps, I'm supportive but happy to defer to others if they don't like it

We could do this here or I could open a separate PR, btw.

**Review by alisterburt — COMMENTED** [review]
Thanks for adding the test! Sorry to be a pain, I actually meant a test like 
https://github.com/napari/napari/blob/0f5281f641ac9ad468648abdc8b3f30b99ae4b70/napari/_qt/layer_controls/_tests/test_qt_image_layer.py#L7-L17
for the Qt stuff added - only being pedantic because we had some problems with the qt controls recently in #3877 (although the issue there is a bit different)

> We could do this here or I could open a separate PR, btw.

Let's go separate PR to keep this one clean and discussion separate!

**Comment by brisvag** [comment]
Added the qt test!

**Review by alisterburt — APPROVED** [review]
Hero! Let's get this in 🥳🥳

**Review by sofroniewn — APPROVED** [review]
Looking through this now. I'm ok to get this in as with the transparency trick as you do things like add the qt button etc which is nice - but it does seem to me like you're not too far off from a better solution.

> Ultimately, I feel like the best way would be to set a slice thickness and simply use that to decide what to display and what not... I'm sure it sounds familiar, @alisterburt :P This actually sounds quite doable on this side, but It's probably supposed to be on the vispy side of things...

I tend to agree that this will be preferable to the transparency behavior

> The way I see it, we should change both of them to use distance <= 0.5; this way, we ensure that they are not lost between slices (as happens for points currently). The downside of this is obviously that we are less true to the data. What do other people think? I'm sure @jni and @sofroniewn have opinions on this :)

I can imagine this being weird. Maybe you can make an issue for this and @grlee77 can also take a look. Or maybe it sounds like you will be working on a new PR for this.

I'm approving feel free to merge tomorrow, but i do think we should follow up with the "thickness" approach instead of the "transparency" approach

**Comment by brisvag** [comment]
> > Ultimately, I feel like the best way would be to set a slice thickness and simply use that to decide what to display and what not... I'm sure it sounds familiar, @alisterburt :P This actually sounds quite doable on this side, but It's probably supposed to be on the vispy side of things...
> 
> I tend to agree that this will be preferable to the transparency behavior

We could implement this using clipping planes under the hood, and exposing a similar API and widget as #3759. This way you can also slice along a different angle, and it's all in "2.5D".

If we want it in "real" 2D slicing, we should probably come up with a generalized solution for a "slice thickness" for all the layers maybe?

> I can imagine this being weird. Maybe you can make an issue for this and @grlee77 can also take a look. Or maybe it sounds like you will be working on a new PR for this.

Yup, the PR with this is at #3997.

----

Btw, the issue with the `nd_vectors` example that @alisterburt mentioned above is fixed now that I added the `abs()` call:

image

**Comment by sofroniewn** [comment]
> If we want it in "real" 2D slicing, we should probably come up with a generalized solution for a "slice thickness" for all the layers maybe?

yeah i like that idea.

I will merge this now as we're only talking about changing implementation behind the scenes. The public API will remain the same.

**Comment by jni** [comment]
Hi folks,

It seems that my suggestion to rename "n_dimensional" to "out_of_slice_display", which @brisvag seemed to agree with, fell by the wayside. Should we open a new PR to rename here and optionally in points?

**Pull request merged**